src/app/file-manager/file-download-modal/file-download-modal.component.ts

```diff
@@ -11,38 +10,172 @@ import { DownloadValidatorService } from '@app/file-manager/services/download-va
     public dialogRef: MatDialogRef<FileDownloadModalComponent>,
     public router: Router,
+    public bucketService: BucketService,
     @Inject(MAT_DIALOG_DATA) public data: any) {
-      if (localStorage.getItem('directory') !== null) {
-        this.directory = localStorage.getItem('directory');
-        this.isValid = true;
+    this.filesMap = new Map();
+
+    if (localStorage.getItem('directory') !== null) {
+      this.directory = localStorage.getItem('directory');
+      this.isValid = true;
+    }
+    BucketService.delimiter = null;
+
+    // this.fileCountObservable.subscribe(
+    //   value => {
+    //     console.log('fileCount: ' + value);
+    //   }
+    // );
+
+    // this.totalSizeObservable.subscribe(
+    //   value => {
+    //     console.log('totalSize: ' + value);
+    //   }
+    // );
+
+  }
+
+  ngOnInit() {
+    this.processFilesToDownload();
+  }
+
+  processFilesToDownload() {
+
+    const patchedItems = this.data.selectedFiles
+      .filter(item => {
+        if (item.id === 'workspaces') {
+          return false;
+        }
+        return true;
+      })
+      .map(
+        item => {
+          if (item.path === '') {
+            return this.cloneItem(item);
+          }
+          return item;
+        }
+      );
+
+    const sortedItems = patchedItems.sort(
+      (itemA, itemB) => {
+        const pathA = itemA.path.toUpperCase();
+        const pathB = itemB.path.toUpperCase();
+        return (pathA < pathB) ? -1 : (pathA > pathB) ? 1 : 0;
+      });
+
+    this.totalFiles = 0;
+
+    this.selectedFiles = [];
+
+    const cleanedItems = this.filterFolderItems(sortedItems);
+
+    this.loadingFiles = true;
+    this.loadingFolders = 0;
+
+    cleanedItems
+      .forEach(
+        item => {
+          if (item.type === 'File') {
+            this.addToSelectedFiles(item, 'directly');
+          } else {
+            // while (this.loadingFolders > 5) {
+            //   // do nothing;
+            //   // this is a dirty, ugly workaround while
+            //   // implementing a better request limiting aproach
+            //   console.log(this.loadingFolders);
+            // }
+            this.getFiles(item);
+          }
+        });
+
+    this.loadingFiles = false;
+  }
+
+  getFiles(folderItem: Item) {
+    let retrievedItems: Item[] = [];
+
+    this.loadingFolders++;
+    this.bucketService.getBucketData(folderItem, null, false, false)
+      .subscribe(
+        elements => {
+          retrievedItems = elements;
+        },
+        error => {
+          console.log('error: ' + JSON.stringify(error, null, 2));
+        },
+        () => {
+          retrievedItems
+            .forEach(
+              child => {
+                this.addToSelectedFiles(child, 'as child');
+              });
+          this.loadingFolders--;
+        }
+      );
+  }
+
+  isLoading() {
+    return this.loadingFiles || this.loadingFolders > 0;
+  }
+
+  addToSelectedFiles(item: Item, mode: string) {
```

Comment:
`mode` is never used.

Reply:
it was used while debugging. Will remove it